pkg/kubelet/kuberuntime/kuberuntime_gc.go

```diff
@@ -356,9 +356,34 @@ func (cgc *containerGC) evictPodLogsDirectories(allSourcesReady bool) error {
 	logSymlinks, _ := osInterface.Glob(filepath.Join(legacyContainerLogsDir, fmt.Sprintf("*.%s", legacyLogSuffix)))
 	for _, logSymlink := range logSymlinks {
 		if _, err := osInterface.Stat(logSymlink); os.IsNotExist(err) {
+			if containerID, err := getContainerIDFromLegacyLogSymlink(logSymlink); err == nil {
+				if status, err := cgc.manager.runtimeService.ContainerStatus(containerID); err == nil {
+					if status.State != runtimeapi.ContainerState_CONTAINER_EXITED {
+						// Here is how container log rotation works (see containerLogManager#rotateLatestLog):
+						//
+						// 1. rename current log to rotated log file whose filename contains current timestamp (fmt.Sprintf("%s.%s", log, timestamp))
+						// 2. reopen the container log
+						// 3. if #2 fails, rename rotated log file back to container log
```

Comment:
Here's how I'd do it:
1. Create blank log file in the same directory as the rename target, with a temporary name.
2. Atomically swap the current log file with the blank log file - Linux has a `RENAME_EXCHANGE` flag to `renameat2()` that could be helpful.

Does that seem like it'd work?
3. Rename the temporary name to the rotated name. If that fails, copy the file contents and unlink the rotated name.

Reply:
Let me think about it.
Normally, with more intermediate files, we need to consider more error cases.

Reply:
On Windows you might need another approach: [`ReplaceFile()`](https://docs.microsoft.com/en-gb/windows/win32/api/winbase/nf-winbase-replacefilea?redirectedfrom=MSDN)

Maybe an idea for a future pull request?

Reply:
Let me search for the golang equivalent.